lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs

```diff
@@ -0,0 +1,907 @@
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( runExceptT )
+import Crypto.Hash
+    ( hash )
+import Data.Bifunctor
+    ( bimap, first )
+import Data.Foldable
+    ( toList )
+import Data.Functor.Classes
+    ( Eq1, Show1 )
+import Data.Map
+    ( Map )
+import Data.Maybe
+    ( catMaybes, fromJust, fromMaybe, isJust, isNothing )
+import Data.Set
+    ( Set )
+import Data.TreeDiff
+    ( ToExpr (..), defaultExprViaShow )
+import GHC.Generics
+    ( Generic )
+import System.Random
+    ( getStdRandom, randomR )
+import Test.QuickCheck
+    ( Arbitrary (..)
+    , Args (..)
+    , Gen
+    , Property
+    , collect
+    , elements
+    , frequency
+    , labelledExamplesWith
+    , property
+    , (===)
+    )
+import Test.QuickCheck.Monadic
+    ( monadicIO )
+import Test.QuickCheck.Random
+    ( mkQCGen )
+import Test.StateMachine
+    ( CommandNames (..)
+    , Concrete
+    , GenSym
+    , Logic (..)
+    , Reason (..)
+    , Reference
+    , StateMachine
+    , Symbolic
+    , elem
+    , forAllCommands
+    , forAllParallelCommands
+    , forall
+    , prettyCommands
+    , prettyParallelCommands
+    , runCommands
+    , runParallelCommands
+    , (.==)
+    )
+import Test.StateMachine.Types
+    ( Command (..), Commands (..), ParallelCommands, ParallelCommandsF (..) )
+
+import qualified Control.Foldl as Foldl
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.List as L
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import qualified Test.QuickCheck as QC
+import qualified Test.StateMachine.Types as QSM
+import qualified Test.StateMachine.Types.Rank2 as Rank2
+
+{-# ANN module ("HLint: ignore Unused LANGUAGE pragma" :: String) #-}
+
+{-------------------------------------------------------------------------------
+  Errors
+-------------------------------------------------------------------------------}
+
+data Err wid
+    = NoSuchWallet wid
+    | WalletAlreadyExists wid
+    deriving (Show, Eq, Functor, Foldable, Traversable)
+
+errNoSuchWallet :: ErrNoSuchWallet -> Err WalletId
+errNoSuchWallet (ErrNoSuchWallet wid) = NoSuchWallet wid
+
+errWalletAlreadyExists :: ErrWalletAlreadyExists -> Err WalletId
+errWalletAlreadyExists (ErrWalletAlreadyExists wid) = WalletAlreadyExists wid
+
+{-------------------------------------------------------------------------------
+  Mock implementation
+-------------------------------------------------------------------------------}
+
+-- | Shortcut for wallet type.
+type MWallet = Wallet (SeqState DummyTarget) DummyTarget
+
+-- | Mock wallet ID -- simple and easy to read
+newtype MWid = MWid String
+    deriving (Show, Eq, Ord, Generic)
+
+widPK :: MWid -> PrimaryKey WalletId
+widPK = PrimaryKey . unMockWid
+
+-- | Convert a mock wallet ID to a real one by hashing it
+unMockWid :: MWid -> WalletId
+unMockWid (MWid wid) = WalletId . hash . B8.pack $ wid
+
+-- | Represent (XPrv, Hash) as a string.
+type MPrivKey = String
```

Comment:
I wonder why we do have this mock `XPrv` and `Hash` in a first place? The generator only takes keys among a fixed list of 3 elements anyway, so we could as well just go for the right type from the start :thinking:  ?